api/src/main/java/io/minio/credentials/AssumeRoleProvider.java

```diff
@@ -0,0 +1,172 @@
+ * API</a>.
+ */
+public class AssumeRoleProvider implements Provider {
+  public static final int DEFAULT_DURATION_SECONDS = (int) TimeUnit.HOURS.toSeconds(1);
+  private final HttpUrl stsEndpoint;
+  private final String accessKey;
+  private final String secretKey;
+  private final String region;
+  private final OkHttpClient httpClient;
+  private final String contentSha256;
+  private final Request request;
+  private Credentials credentials;
+
+  public AssumeRoleProvider(
+      @Nonnull String stsEndpoint,
+      @Nonnull String accessKey,
+      @Nonnull String secretKey,
+      @Nullable Integer durationSeconds,
+      @Nullable String policy,
+      @Nullable String region,
+      @Nullable String roleArn,
+      @Nullable String roleSessionName,
+      @Nullable String externalId,
+      @Nullable OkHttpClient customHttpClient)
+      throws NoSuchAlgorithmException {
+    stsEndpoint = Objects.requireNonNull(stsEndpoint, "STS endpoint cannot be empty");
+    this.stsEndpoint = Objects.requireNonNull(HttpUrl.parse(stsEndpoint), "Invalid STS endpoint");
+    accessKey = Objects.requireNonNull(accessKey, "Access key must not be null");
+    if (accessKey.isEmpty()) {
+      throw new IllegalArgumentException("Access key must not be empty");
+    }
+    this.accessKey = accessKey;
+    this.secretKey = Objects.requireNonNull(secretKey, "Secret key must not be null");
+    this.region = (region != null) ? region : "";
+    this.httpClient = (customHttpClient != null) ? customHttpClient : new OkHttpClient();
+
+    if (externalId != null && (externalId.length() < 2 || externalId.length() > 1224)) {
+      throw new IllegalArgumentException("Length of ExternalId must be in between 2 and 1224");
+    }
+
+    durationSeconds =
+        (durationSeconds != null && durationSeconds > DEFAULT_DURATION_SECONDS)
+            ? durationSeconds
+            : DEFAULT_DURATION_SECONDS;
+
+    HttpUrl.Builder urlBuilder =
+        this.stsEndpoint
+            .newBuilder()
+            .addQueryParameter("Action", "AssumeRole")
+            .addQueryParameter("Version", "2011-06-15")
+            .addQueryParameter("DurationSeconds", String.valueOf(durationSeconds));
+
+    if (roleArn != null) {
+      urlBuilder.addQueryParameter("RoleArn", roleArn);
+    }
+
+    if (roleSessionName != null) {
+      urlBuilder.addQueryParameter("RoleSessionName", roleSessionName);
+    }
+
+    if (policy != null) {
+      urlBuilder.addQueryParameter("Policy", policy);
+    }
+
+    if (externalId != null) {
+      urlBuilder.addQueryParameter("ExternalId", externalId);
+    }
+
+    String data = urlBuilder.build().encodedQuery();
+    this.contentSha256 = Digest.sha256Hash(data);
+    this.request =
+        new Request.Builder()
+            .url(this.stsEndpoint)
+            .method(
+                "POST",
+                RequestBody.create(MediaType.parse("application/x-www-form-urlencoded"), data))
+            .build();
```

Comment:
The request building logic is inside the `fetch()` method in the `WebIdentityClientGrantsProvider` class. Can we try to have a consistent pattern in both? I see a small potential for an abstract base class that provides the `fetch()` logic as well, forcing the concrete sub-classes to build the request, which seems to be the main difference.